### PR TITLE
Firestore `isEqual()` on Public Types

### PR DESCRIPTION
**Pull request by zxu123**
### Discussion
*   See b/62143881
*   Also see inspiration on http://github.com/firebase/firebase-js-sdk/issues/202
### Testing
*   Adds unit test for all relevant `isEqual()`, either already implemented in master or in this branch
*   All unit tests and integration tests pass.
### API Changes
See proposal go/firestoreisequal

### CODE REVIEW

**Comment by mikelehen** [comment]
I think I may be able to get to this before Gil. :-)

File: Firestore/Example/Tests/API/FIRDocumentReferenceTests.m [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  // specific setting as far as we do not fetch data.
+  FIRFirestore *firestore = [[FIRFirestore alloc] initWithProjectID:@"abc"
+                                                           database:@"abc"
+                                                     persistenceKey:@"db123"
+                                                credentialsProvider:nil
+                                                workerDispatchQueue:nil
+                                                        firebaseApp:nil];
+  FSTDocumentKey *keyFoo = [FSTDocumentKey keyWithPathString:@"rooms/foo"];
+  FSTDocumentKey *keyFooDup = [FSTDocumentKey keyWithPathString:@"rooms/foo"];
+  FSTDocumentKey *keyBar = [FSTDocumentKey keyWithPathString:@"rooms/bar"];
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRDocumentReferenceTests.m]: We have a helper: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Util/FSTHelpers.h#L167

File: Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                                                credentialsProvider:nil
+                                                workerDispatchQueue:nil
+                                                        firebaseApp:nil];
+  FSTDocumentKey *keyFoo = [FSTDocumentKey keyWithPathString:@"rooms/foo"];
+  FSTDocumentKey *keyBar = [FSTDocumentKey keyWithPathString:@"rooms/bar"];
+  FSTObjectValue *dateFoo = FSTTestObjectValue(@{ @"a" : @1 });
+  FSTObjectValue *dateBar = FSTTestObjectValue(@{ @"b" : @1 });
+  FSTSnapshotVersion *version = FSTTestVersion(1);
+  FSTDocument *docFoo =
+      [FSTDocument documentWithData:dateFoo key:keyFoo version:version hasLocalMutations:NO];
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m]: Could you use FSTTestDoc() and save some lines?

File: Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  // specific setting as far as we do not fetch data.
+  FIRFirestore *firestore = [[FIRFirestore alloc] initWithProjectID:@"abc"
+                                                           database:@"abc"
+                                                     persistenceKey:@"db123"
+                                                credentialsProvider:nil
+                                                workerDispatchQueue:nil
+                                                        firebaseApp:nil];
+  FSTDocumentKey *keyFoo = [FSTDocumentKey keyWithPathString:@"rooms/foo"];
+  FSTDocumentKey *keyBar = [FSTDocumentKey keyWithPathString:@"rooms/bar"];
+  FSTObjectValue *dateFoo = FSTTestObjectValue(@{ @"a" : @1 });
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m]: date => data I think

File: Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  FSTObjectValue *dateBar = FSTTestObjectValue(@{ @"b" : @1 });
+  FSTSnapshotVersion *version = FSTTestVersion(1);
+  FSTDocument *docFoo =
+      [FSTDocument documentWithData:dateFoo key:keyFoo version:version hasLocalMutations:NO];
+  FSTDocument *docBar =
+      [FSTDocument documentWithData:dateBar key:keyBar version:version hasLocalMutations:NO];
+  XCTAssertEqualObjects([FIRDocumentSnapshot snapshotWithFirestore:firestore
+                                                       documentKey:keyFoo
+                                                          document:nil
+                                                         fromCache:YES],
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRDocumentSnapshotTests.m]: Oh boy. This really turns into a giant wall of text. Can we create a helper to reduce this?

File: Firestore/Example/Tests/API/FIRFieldPathTests.m [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FIRFieldPathTests : XCTestCase
+@end
+
+@implementation FIRFieldPathTests
+
+- (void)testEquals {
+  FSTFieldPath *pathFoo = [FSTFieldPath pathWithServerFormat:@"foo.ooo.oooo"];
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRFieldPathTests.m]: We have a helper for this: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Util/FSTHelpers.h#L158

File: Firestore/Example/Tests/API/FIRQuerySnapshotTests.m [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                                                          snapshot:snapshotBar
+                                                          metadata:metadataFoo]);
+  XCTAssertNotEqualObjects([FIRQuerySnapshot snapshotWithFirestore:firestore
+                                                     originalQuery:queryFoo
+                                                          snapshot:snapshotFoo
+                                                          metadata:metadataFoo],
+                           [FIRQuerySnapshot snapshotWithFirestore:firestore
+                                                     originalQuery:queryFoo
+                                                          snapshot:snapshotFoo
+                                                          metadata:metadataBar]);
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRQuerySnapshotTests.m]: Another giant wall of text.  Can we improve with a helper of some kind?

File: Firestore/Example/Tests/API/FIRQuerySnapshotTests.m [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  // specific setting as far as we do not fetch data.
+  FIRFirestore *firestore = [[FIRFirestore alloc] initWithProjectID:@"abc"
+                                                           database:@"abc"
+                                                     persistenceKey:@"db123"
+                                                credentialsProvider:nil
+                                                workerDispatchQueue:nil
+                                                        firebaseApp:nil];
+  FSTResourcePath *pathFoo = [FSTResourcePath pathWithString:@"foo"];
+  FSTResourcePath *pathBar = [FSTResourcePath pathWithString:@"bar"];
+  FSTQuery *queryFoo = [FSTQuery queryWithPath:pathFoo];
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRQuerySnapshotTests.m]: Helper for this. :-) https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Util/FSTHelpers.h#L196

File: Firestore/Example/Tests/API/FIRQuerySnapshotTests.m (hunk 6 above)
Comment (zxu123) [reply Firestore/Example/Tests/API/FIRQuerySnapshotTests.m]: done

File: Firestore/Source/API/FIRDocumentSnapshot.m [hunk 8]
```diff
@@ -76,6 +76,28 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
...
+  if (self.firestore != snapshot.firestore && ![self.firestore isEqual:snapshot.firestore])
+    return NO;
+  if (self.internalKey != snapshot.internalKey && ![self.internalKey isEqual:snapshot.internalKey])
+    return NO;
+  if (self.internalDocument != snapshot.internalDocument &&
+      (!self.internalDocument || ![self.internalDocument isEqual:snapshot.internalDocument]))
+    return NO;
+  if (self.fromCache != snapshot.fromCache) return NO;
+  return YES;
+}
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Source/API/FIRDocumentSnapshot.m]: done

File: Firestore/Source/API/FIRDocumentSnapshot.m [hunk 9]
```diff
@@ -76,6 +76,28 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
   return self;
 }
 
+// NSObject Methods
+- (BOOL)isEqual:(nullable id)other {
+  if (other == self) return YES;
+  if (!other || ![[other class] isEqual:[self class]]) return NO;
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Source/API/FIRDocumentSnapshot.m]: done

File: Firestore/Source/API/FIRDocumentSnapshot.m [hunk 10]
```diff
@@ -76,6 +76,28 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
...
+
+  return [self isEqualToSnapshot:other];
+}
+
+- (BOOL)isEqualToSnapshot:(nullable FIRDocumentSnapshot *)snapshot {
+  if (self == snapshot) return YES;
+  if (snapshot == nil) return NO;
+  if (self.firestore != snapshot.firestore && ![self.firestore isEqual:snapshot.firestore])
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Source/API/FIRDocumentSnapshot.m]: Let's ignore other properties for now. For the four combination of .firestore.

*   case 01) self.firestore == nil, snapshot.firestore != nil, should return NO;
*   case 00) self.firestore == nil, snapshot.firestore == nil, should return YES;
*   case 11) self.firestore != nil, snapshot.firestore != nil, should return whether self.firestore == snapshot.firestore;
*   case 10) self.firestore != nil, snapshot.firestore == nil, should return NO;

As far as self.firestore != nil and FIRFirestore:isEqual implemented correctly, [self.firestore isEqual:snapshot.firestore] is enough to get the result;
As soon as self.firestore == nil, [self.firestore isEqual:snapshot.firestore] will always returns nil and we cannot tell apart 01 and 00 cases. Additional self.firestore != snapshot.firestore helps in this case. (and as side effect, may or may not improve performance by checking reference.)

File: Firestore/Source/API/FIRFieldPath.m [hunk 11]
```diff
@@ -80,12 +80,12 @@ - (id)copyWithZone:(NSZone *__nullable)zone {
...
 }
 
-- (BOOL)isEqual:(id)object {
+- (BOOL)isEqual:(nullable id)object {
   if (self == object) {
     return YES;
   }
 
-  if (![object isKindOfClass:[FIRFieldPath class]]) {
+  if (!object || ![object isKindOfClass:[FIRFieldPath class]]) {
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Source/API/FIRFieldPath.m]: done

File: Firestore/Source/API/FIRDocumentSnapshot.m [hunk 12]
```diff
@@ -76,6 +76,37 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
...
+    return NO;
+  if (self.internalKey != snapshot.internalKey && ![self.internalKey isEqual:snapshot.internalKey])
+    return NO;
+  if (self.internalDocument != snapshot.internalDocument &&
+      (!self.internalDocument || ![self.internalDocument isEqual:snapshot.internalDocument]))
```
Comment (mikelehen) [inline Firestore/Source/API/FIRDocumentSnapshot.m]: You don't need the !self.internalDocument check do you? nil inequality will be caught by the first check as per your explanation above. :-)

File: Firestore/Source/API/FIRQuerySnapshot.m [hunk 13]
```diff
@@ -76,6 +77,35 @@ - (instancetype)initWithFirestore:(FIRFirestore *)firestore
...
+  if (other == self) return YES;
+  if (![[other class] isEqual:[self class]]) return NO;
+
+  return [self isEqualToSnapshot:other];
+}
+
+- (BOOL)isEqualToSnapshot:(nullable FIRQuerySnapshot *)snapshot {
+  if (self == snapshot) return YES;
+  if (snapshot == nil) return NO;
+  if (self.firestore != snapshot.firestore && ![self.firestore isEqual:snapshot.firestore])
```
Comment (mikelehen) [inline Firestore/Source/API/FIRQuerySnapshot.m]: shouldn't these be `||`s ?

File: Firestore/Source/API/FIRQuery.m [hunk 14]
```diff
@@ -107,7 +107,7 @@ - (instancetype)initWithQuery:(FSTQuery *)query firestore:(FIRFirestore *)firest
 
 - (BOOL)isEqual:(nullable id)other {
   if (other == self) return YES;
-  if (!other || ![[other class] isEqual:[self class]]) return NO;
+  if (![[other class] isEqual:[self class]]) return NO;
```
Comment (mikelehen) [inline Firestore/Source/API/FIRQuery.m]: If you haven't already, can you just double-check that this still works?  I'm a tiny bit nervous because AppCode auto-generates these methods with the "!other" check, and I don't know why.

File: Firestore/Source/Core/FSTView.m [hunk 15]
```diff
@@ -94,6 +94,12 @@ - (BOOL)isEqual:(id)other {
   return self.type == otherChange.type && [self.key isEqual:otherChange.key];
 }
 
+- (NSUInteger)hash {
+  NSUInteger hash = self.type;
+  hash = hash * 31u + [self.key hash];
+  return hash;
+}
```
Comment (mikelehen) [inline Firestore/Source/Core/FSTView.m]: Ooh! Good catch.

File: Firestore/Example/Tests/Util/FSTHelpers.h [hunk 16]
```diff
@@ -181,6 +187,14 @@ FSTDocument *FSTTestDoc(NSString *path,
                         NSDictionary<NSString *, id> *data,
                         BOOL hasMutations);
 
+/** A convenience method for creating a doc snapshot for tests. */
+FIRDocumentSnapshot *FSTTestDocSnapshot(
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Example/Tests/Util/FSTHelpers.h]: done. refactored and added `FSTAPIHelpers.h/m` which contains all helpers with `FIR*` return type.

File: Firestore/Example/Tests/Util/FSTHelpers.h [hunk 17]
```diff
@@ -195,6 +209,14 @@ FSTDocumentKeyReference *FSTTestRef(NSString *projectID, NSString *databaseID, N
 /** A convenience method for creating a query for the given path (without any other filters). */
 FSTQuery *FSTTestQuery(NSString *path);
 
+/** A convenience method for creating a particular query snapshots for tests. */
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Example/Tests/Util/FSTHelpers.h]: done

File: Firestore/Example/Tests/Util/FSTHelpers.m [hunk 18]
```diff
@@ -171,6 +208,40 @@
...
+        NSString *path,
+        NSArray<NSDictionary<NSString *, id> *> *oldData,
+        NSArray<NSDictionary<NSString *, id> *> *dataToAdd,
+        BOOL hasPendingWrites,
+        BOOL fromCache) {
+  FIRSnapshotMetadata *metadata = [FIRSnapshotMetadata snapshotMetadataWithPendingWrites:hasPendingWrites
+                                                                               fromCache:fromCache];
+  FSTDocumentSet *oldDocuments = FSTTestDocSet(FSTDocumentComparatorByKey, @[]);
+  for (NSDictionary<NSString *, id> *data in oldData) {
+    oldDocuments = [oldDocuments documentSetByAddingDocument:FSTTestDoc(path, 1, data, hasPendingWrites)];
```
[the remark this answers is not in the document]
Comment (zxu123) [reply Firestore/Example/Tests/Util/FSTHelpers.m]: done

File: Firestore/Source/API/FIRQuery.m (hunk 14 above)
Comment (zxu123) [reply Firestore/Source/API/FIRQuery.m]: Still works, see logic given below:

The only difference happens when `other` is `nil`. That case one return `NO` immediately while the other will check `![[other class] isEqual:[self class]]`, which is equivalent to `![nil isEqual:[self class]]`, which is equivalent to `!nil`, which makes the method return `NO` as well. So `!other` check is redundant.

I have no particular preference on one over the other. So I can change it back if you think with `!other` check is easier to understand.

File: Firestore/Source/API/FIRDocumentSnapshot.m (hunk 12 above)
Comment (zxu123) [reply Firestore/Source/API/FIRDocumentSnapshot.m]: Yeah, the check is redundant. Removed.

File: Firestore/Source/API/FIRQuerySnapshot.m (hunk 13 above)
Comment (zxu123) [reply Firestore/Source/API/FIRQuerySnapshot.m]: No, should be &&. Check the case-analysis above (nil v.s. non-nil, non-nil v.s. non-nil). Another way to look at this is to use double-negation of it. Say, self.firestore is equal iff the reference is equal OR the content is equal. So self.firestore is not equal (case here to return NO) iff the reference is not equal AND the content is not equal.

File: Firestore/Example/Tests/API/FSTAPIHelpers.m [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+#import "Firestore/Source/Model/FSTDatabaseID.h"
+#import "Firestore/Source/Model/FSTDocument.h"
+#import "Firestore/Source/Model/FSTDocumentKey.h"
+#import "Firestore/Source/Model/FSTDocumentSet.h"
+#import "Firestore/Source/Model/FSTFieldValue.h"
+#import "Firestore/Source/Model/FSTMutation.h"
+#import "Firestore/Source/Model/FSTPath.h"
+#import "Firestore/Source/Remote/FSTRemoteEvent.h"
+#import "Firestore/Source/Remote/FSTWatchChange.h"
+#import "Firestore/Source/Util/FSTAssert.h"
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.m]: Most of these imports are unused.  Per AppCode, this is the set you need:
```
#import "Firestore/Example/Tests/API/FSTAPIHelpers.h"
#import <Firestore/Source/API/FIRDocumentSnapshot+Internal.h>
#import <Firestore/Source/API/FIRQuerySnapshot+Internal.h>
#import <Firestore/Source/API/FIRSnapshotMetadata+Internal.h>
#import <Firestore/Source/Core/FSTViewSnapshot.h>

#import "FirebaseFirestore/FIRCollectionReference.h"
#import "FirebaseFirestore/FIRDocumentReference.h"
#import "FirebaseFirestore/FIRDocumentSnapshot.h"
#import "FirebaseFirestore/FIRFirestore.h"
#import "FirebaseFirestore/FIRGeoPoint.h"
#import "FirebaseFirestore/FIRQuerySnapshot.h"
#import "FirebaseFirestore/FIRSnapshotMetadata.h"
#import "Firestore/Source/API/FIRCollectionReference+Internal.h"
#import "Firestore/Source/API/FIRFirestore+Internal.h"
#import "Firestore/Source/Core/FSTQuery.h"
#import "Firestore/Source/Model/FSTDocument.h"
#import "Firestore/Source/Model/FSTDocumentKey.h"
#import "Firestore/Source/Model/FSTDocumentSet.h"
#import "Firestore/Source/Model/FSTPath.h"
```

File: Firestore/Example/Tests/API/FIRGeoPointTests.m [hunk 20]
```diff
@@ -28,27 +29,24 @@ @interface FIRGeoPointTests : XCTestCase
...
-  XCTAssertNotEqualObjects([[FIRGeoPoint alloc] initWithLatitude:0 longitude:0],
-                           [[FIRGeoPoint alloc] initWithLatitude:1 longitude:0]);
-  XCTAssertNotEqualObjects([[FIRGeoPoint alloc] initWithLatitude:0 longitude:0],
-                           [[FIRGeoPoint alloc] initWithLatitude:0 longitude:1]);
-  XCTAssertNotEqualObjects([[FIRGeoPoint alloc] initWithLatitude:0 longitude:0],
-                           [[NSObject alloc] init]);
+  FIRGeoPoint *foo = FSTTestGeoPoint(0, 0);
+  FIRGeoPoint *fooDup = FSTTestGeoPoint(0, 0);
+  FIRGeoPoint *bar = FSTTestGeoPoint(1.23, 4.56);
+  FIRGeoPoint *barDup = FSTTestGeoPoint(1.23, 4.56);
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRGeoPointTests.m]: I'm not sure bar / barDup is adding anything to these tests.  You do the same equals checks as with foo / fooDup. If you want non-zero values, I'd just change foo to be non-zero values.

File: Firestore/Example/Tests/API/FIRGeoPointTests.m [hunk 21]
```diff
@@ -28,27 +29,24 @@ @interface FIRGeoPointTests : XCTestCase
...
-          [[[FIRGeoPoint alloc] initWithLatitude:1 longitude:0] hash]);
-  XCTAssertNotEqual([[[FIRGeoPoint alloc] initWithLatitude:0 longitude:0] hash],
-          [[[FIRGeoPoint alloc] initWithLatitude:0 longitude:1] hash]);
-  XCTAssertNotEqual([[[FIRGeoPoint alloc] initWithLatitude:0 longitude:0] hash],
-          [[[NSObject alloc] init] hash]);
+  XCTAssertEqual([foo hash], [fooDup hash]);
+  XCTAssertEqual([bar hash], [barDup hash]);
+  XCTAssertNotEqual([foo hash], [differentLatitude hash]);
+  XCTAssertNotEqual([foo hash], [differentLongitude hash]);
+  XCTAssertNotEqual([foo hash], [object hash]);
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FIRGeoPointTests.m]: Why did you add the object hash test for this test but not others?  (I would drop it in general. I'm not sure what it's protecting against)

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+@class FSTTransformMutation;
+@class FSTView;
+@class FSTViewSnapshot;
+@class FSTObjectValue;
+@protocol FSTFilter;
+
+NS_ASSUME_NONNULL_BEGIN
+
+#if __cplusplus
+extern "C" {
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: I don't believe this is required (usually you need this when you're going to be including the header file from C/C++ code).

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * A convenience method for creating a particular query snapshot for tests.
+ * This function allows user to pass in snapshot of the query in the past as well as new rows to be
+ * added into the snapshot as for now. The current snapshot of the query consists both data.
+ */
+FIRQuerySnapshot *FSTTestQuerySnapshot(
+    NSString *path,
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *oldData,
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *dataToAdd,
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: Can you rename these `oldDocs` and `docsToAdd`?

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+/** A covenience method for creating a collection reference from a path string. */
+FIRCollectionReference *FSTTestCollectionRef(NSString *path);
+
+/** A covenience method for creating a document reference from a path string. */
+FIRDocumentReference *FSTTestDocRef(NSString *path);
+
+/**
+ * A convenience method for creating a particular query snapshot for tests.
+ * This function allows user to pass in snapshot of the query in the past as well as new rows to be
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: What are "rows"? :-)  [Firestore deals with collections and documents, not tables and rows]

I'd also convert this paragraph into `@param` comments.  (I'm fine with omitting `@param` comments when the parameters are straightforward, but if you're going to explain them it's probably better to do it in the official way than via prose)

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+ * added into the snapshot as for now. The current snapshot of the query consists both data.
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: "as for now" => "as of now", "consists both data" => "consists of both sets of documents"  and maybe turn this into a `@returns` comment.

File: Firestore/Example/Tests/Model/FSTFieldValueTests.m [hunk 26]
```diff
@@ -26,6 +26,7 @@
 #import "Firestore/Source/Model/FSTFieldValue.h"
 #import "Firestore/Source/Model/FSTPath.h"
 
+#import "Firestore/Example/Tests/API/FSTAPIHelpers.h"
```
Comment (mikelehen) [inline Firestore/Example/Tests/Model/FSTFieldValueTests.m]: Hrm...  I didn't realize that these tests would end up relying on FSTAPIHelpers.h.  That makes it less obvious that my suggested separation makes sense. :-/

I think the difficulty comes from the fact that FIRGeoPoint is used as both a "model" class and an "API" class. I'd suggest moving FSTTestGeoPoint back into FSTHelpers.h/.m and adding a comment something like:

// Note that FIRGeoPoint is a model class in addition to an API class, so we put this helper here instead of FSTAPIHelpers.h

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+@class FSTSetMutation;
+@class FSTSnapshotVersion;
+@class FSTSortOrder;
+@class FSTTargetChange;
+@class FSTTimestamp;
+@class FSTTransformMutation;
+@class FSTView;
+@class FSTViewSnapshot;
+@class FSTObjectValue;
+@protocol FSTFilter;
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: The vast majority of these are unused.  Unfortunately AppCode doesn't seem to tell me which ones.  Can you take a manual pass through cleaning these up (e.g. delete them all and then re-add as necessary)

File: Firestore/Example/Tests/Util/FSTHelpers.h [hunk 28]
```diff
@@ -21,7 +21,12 @@
 #import "Firestore/Source/Model/FSTDocumentDictionary.h"
 #import "Firestore/Source/Model/FSTDocumentKeySet.h"
 
+@class FIRCollectionReference;
+@class FIRDocumentReference;
+@class FIRDocumentSnapshot;
+@class FIRFirestore;
 @class FIRGeoPoint;
+@class FIRQuerySnapshot;
```
Comment (mikelehen) [inline Firestore/Example/Tests/Util/FSTHelpers.h]: I think all of these added FIR* forward declarations are no longer necessary?

File: Firestore/Example/Tests/Util/FSTHelpers.m [hunk 29]
```diff
@@ -14,18 +14,22 @@
...
+#import "FirebaseFirestore/FIRCollectionReference.h"
+#import "FirebaseFirestore/FIRDocumentReference.h"
 #import "FirebaseFirestore/FIRDocumentSnapshot.h"
 #import "FirebaseFirestore/FIRFieldPath.h"
 #import "FirebaseFirestore/FIRFirestore.h"
 #import "FirebaseFirestore/FIRGeoPoint.h"
 #import "FirebaseFirestore/FIRQuerySnapshot.h"
 #import "FirebaseFirestore/FIRSnapshotMetadata.h"
+#import "Firestore/Source/API/FIRCollectionReference+Internal.h"
+#import "Firestore/Source/API/FIRDocumentReference+Internal.h"
```
Comment (mikelehen) [inline Firestore/Example/Tests/Util/FSTHelpers.m]: I think the added imports are no longer necessary?

File: Firestore/Source/API/FIRQuerySnapshot.m (hunk 13 above)
Comment (mikelehen) [reply Firestore/Source/API/FIRQuerySnapshot.m]: Er, sorry I keep trying to lead you astray.  For some reason when I was looking at this I was thinking of the equals case instead of the not-equals case. Carry on!

File: Firestore/Example/Tests/API/FSTAPIHelpers.h (hunk 22 above)
Comment (wilhuff) [reply Firestore/Example/Tests/API/FSTAPIHelpers.h]: For the most part this is going to be required in every Objective-C header as we go so we might as well put these in. It's not worth thinking about whether or not they're required.

File: Firestore/Example/Tests/API/FSTAPIHelpers.h (hunk 27 above)
Comment (zxu123) [reply Firestore/Example/Tests/API/FSTAPIHelpers.h]: done

File: Firestore/Example/Tests/API/FSTAPIHelpers.m (hunk 19 above)
Comment (zxu123) [reply Firestore/Example/Tests/API/FSTAPIHelpers.m]: done, also clean-up more and make style consistent (<> vs "")

File: Firestore/Example/Tests/Model/FSTFieldValueTests.m (hunk 26 above)
Comment (zxu123) [reply Firestore/Example/Tests/Model/FSTFieldValueTests.m]: done

File: Firestore/Example/Tests/Util/FSTHelpers.h (hunk 28 above)
Comment (zxu123) [reply Firestore/Example/Tests/Util/FSTHelpers.h]: done

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 30]
```diff
@@ -62,31 +32,35 @@ extern "C" {
...
+ * @param hasPendingWrites Whether the query snapshot has pending writes to the server.
+ * @param fromCache Whether the query snapshot is cache result.
+ * @returns A query snapshot that consists of both sets of documents.
  */
 FIRQuerySnapshot *FSTTestQuerySnapshot(
     NSString *path,
-    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *oldData,
-    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *dataToAdd,
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *oldDocs,
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *DocsToAdd,
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: lowercase please. :-)  (DocsToAdd=>docsToAdd)

File: Firestore/Example/Tests/API/FSTAPIHelpers.h [hunk 31]
```diff
@@ -62,31 +32,35 @@ extern "C" {
...
 FIRDocumentReference *FSTTestDocRef(NSString *path);
 
 /**
  * A convenience method for creating a particular query snapshot for tests.
- * This function allows user to pass in snapshot of the query in the past as well as new rows to be
- * added into the snapshot as for now. The current snapshot of the query consists both data.
+ *
+ * @param path To be used in constructing the query.
+ * @param oldDocs Provides data to construct the query snapshot in the past. It maps each key to a
+ * document. The key is the document's path relative to the query path.
```
Comment (mikelehen) [inline Firestore/Example/Tests/API/FSTAPIHelpers.h]: nit: "document's path" suggests this could be deeply nested under the query path, but it really must only be a single level deep. I would say:

Provides the prior set of documents in the QuerySnapshot. Each dictionary entry maps to a document, with the key being the document id, and the value being the document contents.
